hack/windows/Install-OVS.ps1

```diff
@@ -187,33 +118,272 @@ function InstallOVS() {
+    $mininalVersion = [version]$VCRedistsVersion
+    $existingVCRedists = getInstalledVcRedists
+    foreach ($redist in $existingVCRedists) {
+        $installedVersion = [version]$redist.Version
+        # VC redists files with a higher version are installed, return.
+        if ($installedVersion -ge $mininalVersion) {
+            return
+        }
     }
-    Log "Hyper-V Virtual Machine Management service status: $VMMSStatus"
-    if ($VMMSStatus -eq "Running") {
-        cmd /c "cd $OVSDriverDir && install.cmd"
-    } else {
-        cd $OVSDriverDir ; netcfg -l .\ovsext.inf -c s -i OVSExt; cd $WorkDir
+    if (-not (Test-Path $VCRedistPath)) {
+        mkdir -p $VCRedistPath
+        curl.exe -Lo $VCRedistPath\vc_redist.x64.exe $DefaultVCRedistsDownloadURL
+    }
+    # Install the VC redistributable files.
+    Get-ChildItem $VCRedistPath -Filter *.exe | ForEach-Object {
+        Start-Process -FilePath $_.FullName -Args '/install /passive /norestart' -Verb RunAs -Wait
+    }
+}
+
+function CheckAndInstallOVSDriver {
+    param (
+        [Parameter(Mandatory = $true)]
+        [String]$OVSDriverPath
+    )
+
+    $expVersion = [version]$(Get-Item $OVSDriverPath\ovsext.sys).VersionInfo.ProductVersion
+    $ovsInstalled = $(netcfg -q ovsext) -like "*is installed*"
+    $installedDrivers = getInstalledOVSDrivers
+
+    # OVSext driver with the desired version is already installed, return
+    if ($ovsInstalled -and (@($installedDrivers).Length -eq 1) -and ($installedDrivers[0].DriverVersion -eq $expVersion)){
+        return
+    }
+
+    $global:updateOVS = $true
+    # Uninstall the existing driver which is with a different version.
+    if ($ovsInstalled) {
+        netcfg -u ovsext
+    }
+
+    # Clean up the installed ovsext drivers packages.
+    foreach ($driver in $installedDrivers) {
+        $publishdName = $driver.PublishedName
+        pnputil.exe -d $publishdName
+    }
+
+    # Import OVSext driver certificate to TrustedPublisher and Root.
+    $DriverFile="$OVSDriverPath\ovsext.sys"
+    $CertificateFile = "$OVSDriverPath\package.cer"
+    $ExportType = [System.Security.Cryptography.X509Certificates.X509ContentType]::Cert
+    $Cert = (Get-AuthenticodeSignature $DriverFile).SignerCertificate
+    [System.IO.File]::WriteAllBytes($CertificateFile, $Cert.Export($ExportType))
+    Import-Certificate -FilePath "$CertificateFile" -CertStoreLocation cert:\LocalMachine\TrustedPublisher
+    Import-Certificate -FilePath "$CertificateFile" -CertStoreLocation cert:\LocalMachine\Root
+
+    # Install the OVSext driver with the desired version
+    # Copy $OVSDriverPath to host path "c:/openvwitch/driver", and then install ovsext.inf with the host path.
+    # This is a workaround for error code "0x80070003" with containerd v1.7+. The error happens when Windows
+    #  utitilty netcfg tries to access the container's mount path "C:/hpc/".
+    $hostDriverPath="${OVSInstallDir}\driver"
+    if ($OVSDriverPath -ne $hostDriverPath){
```

Comment:
I think we should remove this test in _all_ the copy functions.
I cannot think of any remaining legitimate case in which `$OVSLocalPath` and `$OVSInstallDir` would be the same? In order for this to happen, the user would have to invoke the script with `LocalFile == OVSInstallDir`, which IMO does not need to be a supported case (we could exit with an error). Maybe I am missing something though.

Reply:
For the use scenario, sometimes we may need to replace OVS driver on the Windows host for test purpose, e.g., adding logs in the ovsext driver to perform troubleshooting sessions, at this time we can directly replace the driver files in `$hostDriverPath`, and then call the script. 

This is not for production usage, so if you prefer we ignore it, I can remove the compare and this check in the start,

```
if (($LocalFile -ne "") -and ("$LocalFile" -eq "$OVSInstallDir")) {
    Log "LocalFile and OVSInstallDir are the same, exiting"
    exit 1
}
```

Reply:
I see. But for troubleshooting wouldn't you also need to somehow overwrite the OVS driver version, or `CheckAndInstallOVSDriver` would return early?

If this is convenient for debugging and there is no simple alternative, then I think you should keep it, and maybe just log a warning when `"$LocalFile" -eq "$OVSInstallDir"`.

Reply:
In my last change, I remove the path compare on driver path and usr/bin as you suggested, and add a precheck at the start to reject the input if `$LocalFile` equals to `$OVSInstallDir`. For the local troubleshooting (not production env), I can call UnInstall-OVS.ps1 then call Install-OVS.ps1 as an alternative. Please help review the patch again. @antoninbas